lib/elixir/lib/module/parallel_checker.ex

```diff
@@ -0,0 +1,268 @@
+defmodule Module.ParallelChecker do
+  @moduledoc false
+
+  @type cache() :: {pid(), :ets.tid()}
+  @type warning() :: term()
+  @type kind() :: :def | :defmacro
+
+  @doc """
+  Receives pairs of module maps and BEAM binaries. In parallel it verifies
+  the modules and adds the ExCk chunk to the binaries. Returns the updated
+  binaries and a list of warnings from the verification.
+  """
+  @spec verify([%{}], pos_integer()) :: {[{module(), binary()}], [warning()]}
+  def verify([], _schedulers) do
+    {[], []}
+  end
+
+  def verify(modules, schedulers) do
+    {:ok, server} = :gen_server.start_link(__MODULE__, [modules, schedulers], [])
+
+    try do
+      start_and_wait(server)
+    after
+      :gen_server.stop(server)
+    end
+  end
+
+  @doc """
+  Preloads a module into the cache. Call this function before any other
+  cache lookups for the module.
+  """
+  @spec preload_module(cache(), module()) :: :ok
+  def preload_module({server, ets}, module) do
+    case :ets.lookup(ets, {:cached, module}) do
+      [{_key, _}] -> :ok
+      [] -> cache_module({server, ets}, module)
+    end
+  end
+
+  @doc """
+  Returns the export kind for the given MFA from the cache. If the module
+  does not exist return `{:error, :module}`, or if the function does not
+  exist return `{:error, :function}`.
+  """
+  @spec fetch_export(cache(), module(), atom(), arity()) ::
+          {:ok, kind()} | {:error, :function | :module}
+  def fetch_export({_server, ets}, module, fun, arity) do
+    case :ets.lookup(ets, {:cached, module}) do
+      [{_key, true}] ->
+        case :ets.lookup(ets, {:export, {module, fun, arity}}) do
+          [{_key, kind, _reason}] -> {:ok, kind}
```

Comment:
Given `reason` is already available here, I would make it return `{:ok, kind, reason}` and already check the reason in the caller.